sync/src/synchronizer/block_fetcher.rs

```diff
@@ -123,26 +123,42 @@ impl BlockFetcher {
                     break;
                 }
 
-                let to_fetch = self
-                    .snapshot
-                    .get_ancestor(&best_known_header.hash(), index_height)?;
-
-                // NOTE: Filtering `BLOCK_STORED` but not `BLOCK_RECEIVED`, is for avoiding
-                // stopping synchronization even when orphan_pool maintains dirty items by bugs.
-                if self
-                    .snapshot
-                    .contains_block_status(&to_fetch.hash(), BlockStatus::BLOCK_STORED)
-                {
-                    continue;
-                }
-
-                if inflight.insert(self.peer, to_fetch.hash()) {
-                    trace!(
-                        "[Synchronizer] inflight insert {:?}------------{}",
-                        to_fetch.number(),
-                        to_fetch.hash(),
-                    );
-                    fetch.push(to_fetch.hash());
+                // Try to get [from, to) headers, return the corresponding headers in reversed order.
+                let from = index_height;
+                let to = from + (count - fetch.len()) as u64;
+                let candidates = {
+                    let mut candidates = Vec::with_capacity((to - from) as usize);
+                    let mut parent_hash = self
+                        .snapshot
+                        .get_ancestor(&best_known_header.hash(), to)?
+                        .parent_hash();
+                    for _ in from..to {
+                        let header = self.snapshot.get_header_view(&parent_hash)?;
+                        parent_hash = header.parent_hash();
+                        candidates.push(header.into_inner());
+                    }
+                    candidates
+                };
+                index_height = to - 1;
+
+                for to_fetch in candidates.into_iter().rev() {
+                    // NOTE: Filtering `BLOCK_STORED` but not `BLOCK_RECEIVED`, is for avoiding
+                    // stopping synchronization even when orphan_pool maintains dirty items by bugs.
+                    if self
+                        .snapshot
+                        .contains_block_status(&to_fetch.hash(), BlockStatus::BLOCK_STORED)
+                    {
+                        continue;
+                    }
+
+                    if inflight.insert(self.peer, to_fetch.hash()) {
+                        trace!(
+                            "[Synchronizer] inflight insert {:?}------------{}",
+                            to_fetch.number(),
+                            to_fetch.hash(),
+                        );
+                        fetch.push(to_fetch.hash());
+                    }
```

Comment:
Why two level of loop?

Reply:
For example, when we search block[100..116], the original code will involve `16` times of `get_ancestor`, `get_ancestor(100), get_ancestor(101), ... get_ancestor(105)`.

And now changed to: 
(1) `get_ancestor(16)`
(2) get header[100..115] via parent tracing.

That will save 15 times of `get_ancestor` involving.

Reply:
I think `get_ancestor` should be a cheap action, when it's not a hot code, better with simple code. I worry about the two level loop introduce new bug.